src/Razor/src/Microsoft.AspNetCore.Razor.VSCode/src/RazorDefinitionProvider.ts

```diff
@@ -0,0 +1,39 @@
+/* --------------------------------------------------------------------------------------------
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See License.txt in the project root for license information.
+ * ------------------------------------------------------------------------------------------ */
+
+import * as vscode from 'vscode';
+import { RazorLanguageFeatureBase } from './RazorLanguageFeatureBase';
+
+export class RazorDefinitionProvider
+    extends RazorLanguageFeatureBase
+    implements vscode.DefinitionProvider {
+
+    public async provideDefinition(
+        document: vscode.TextDocument, position: vscode.Position,
+        token: vscode.CancellationToken) {
+
+        const projection = await this.getProjection(document, position, token);
+        if (projection) {
+            const definitions = await vscode.commands.executeCommand<vscode.Definition>(
+                'vscode.executeDefinitionProvider',
+                projection.uri,
+                projection.position);
+            // C# knows about line pragma, if we're getting a direction to a virtual c# document
+            // that means the piece we're trying to navigate to does not have a representation in the
+            // top level file.
+            const result = (definitions as vscode.Location[]).filter(element => {
+                return !(element.uri.path.endsWith('__virtual.cs'));
+            });
+
+            result.forEach(element => {
+                const uri = element.uri.fsPath || element.uri.path;
+                const path = uri.toString().replace('.cshtml__virtual.html', '.cshtml');
```

Comment:
at this point "uri" is already a string isn't it? If so I'd suggest renaming `uri` to `uriPath`. Also i'd add a comment here stating why this is necessary, similarly to your above mention of C# virutal document navigations.

Reply:
Also, this wont work for .razor files, I'd change the suffix comparison to be less concrete, aka `replace('__virtual.html', '')`